.github/workflows/test.yml

```diff
@@ -0,0 +1,174 @@
+  # the base. The container image is passed between the jobs as an artifact.
+  test-vr-xcon:
+    runs-on: ubuntu-22.04
+    env:
+      TOOL: vr-xcon
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Build ${{ env.TOOL }}
+        run: |
+          make ${{ env.TOOL }}
+
+      - name: Test ${{ env.TOOL }}
+        run: |
+          make ${{ env.TOOL }}-test
+
+      - name: Export ${{ env.TOOL }} container image
+        run: docker save -o vr-xcon.tar ${DOCKER_REGISTRY}/vr-xcon
+
+      - name: Upload ${{ env.TOOL }} container image artifact
+        uses: actions/upload-artifact@v3
+        with:
+          name: vr-xcon
+          path: vr-xcon.tar
+
+      - name: Save ${{ env.TOOL }} logs
+        if: always()
+        run: |
+          make -C ${{ env.TOOL }} docker-test-save-logs
+
+      - uses: actions/upload-artifact@v3
+        if: always()
+        with:
+          name: tool-logs
+          path: |
+            ${{ env.TOOL }}/*.log
+
+      - name: Clean test ${{ env.TOOL }} containers
+        if: always()
+        run: make -C ${{ env.TOOL }} docker-test-clean
+  test-vr-bgp:
+    runs-on: ubuntu-22.04
+    needs: test-vr-xcon
+    env:
+      TOOL: vr-bgp
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Download vr-xcon container image artifact
+        uses: actions/download-artifact@v3
+        with:
+          name: vr-xcon
+
+      - name: Import vr-xcon container image
+        run: docker load -i vr-xcon.tar
+
+      - name: Build ${{ env.TOOL }}
+        run: |
+          make ${{ env.TOOL }}
+
+      - name: Test ${{ env.TOOL }}
+        run: |
+          make ${{ env.TOOL }}-test
+
+      - name: Save ${{ env.TOOL }} logs
+        if: always()
+        run: |
+          make -C ${{ env.TOOL }} docker-test-save-logs
+
+      - uses: actions/upload-artifact@v3
+        if: always()
+        with:
+          name: tool-logs
+          path: |
+            ${{ env.TOOL }}/*.log
+
+      - name: Clean test ${{ env.TOOL }} containers
+        if: always()
+        run: make -C ${{ env.TOOL }} docker-test-clean
+
+  test-vr:
+    runs-on: ["self-hosted"]
+    container:
+      image: vrnetlab/ci-builder
+      volumes:
+        - /data/gh-runner/vrnetlab-images:/vrnetlab-images
+    strategy:
+      fail-fast: false
+      matrix:
+        platform: ['csr', 'nxos', 'routeros', 'sros', 'veos', 'vmx', 'vqfx', 'vrp', 'vsr1000', 'xrv', 'xrv9k']
+
+    steps:
+      - uses: actions/checkout@v3
+      - name: Fixup dubious ownership
+        run: git config --global --add safe.directory ${GITHUB_WORKSPACE}
+
+      - name: Use git to check if the source files (platform or shared) have changed
+        uses: dorny/paths-filter@v2
+        id: source_changes
+        with:
+          filters: |
+            platform:
+              - '${{ matrix.platform }}/**/*'
+              - 'common/*'
+
+      - name: Compute platform image hashes in bind-mounted volume
```

Comment:
I'd really prefer if a lot of these things were in a Make target instead... it's so hard to develop in Github Actions. But since it's already written, nevermind... just try to keep as much as possible in Make or similar to enable running & testing locally :)

Reply:
Next time :)

These scripts were added specifically for GitHub Actions since GitLab + LFS has the equivalent change detection and per-job rules built-in.